aloha-core/src/main/scala/com/eharmony/aloha/dataset/vw/multilabel/VwDownsampledMultilabelRowCreator.scala

```diff
@@ -0,0 +1,378 @@
+  * @tparam K the label or class type
+  * @author deaktator
+  * @since 11/6/2017
+  */
+final case class VwDownsampledMultilabelRowCreator[-A, K](
+    allLabelsInTrainingSet: sci.IndexedSeq[K],
+    featuresFunction: FeatureExtractorFunction[A, Sparse],
+    defaultNamespace: List[Int],
+    namespaces: List[(String, List[Int])],
+    normalizer: Option[CharSequence => CharSequence],
+    positiveLabelsFunction: GenAggFunc[A, sci.IndexedSeq[K]],
+    classNs: Char,
+    dummyClassNs: Char,
+    numDownsampledNegLabels: Int,
+    seedCreator: () => Long,
+    includeZeroValues: Boolean = false
+) extends StatefulRowCreator[A, Array[String], Long]
+     with Logging {
+
+  require(
+    0 < numDownsampledNegLabels,
+    s"numDownsampledNegLabels must be positive, found $numDownsampledNegLabels"
+  )
+
+  import VwDownsampledMultilabelRowCreator._
+
+  @transient private[this] lazy val labelToInd = allLabelsInTrainingSet.zipWithIndex.toMap
+
+  // Precomputed for efficiency.
+
+  private[this] val negativeDummyStr =
+    s"$NegDummyClassId:$NegativeCost |$dummyClassNs $NegativeDummyClassFeature"
+
+  private[this] val positiveDummyStr =
+    s"$PosDummyClassId:$PositiveCost |$dummyClassNs $PositiveDummyClassFeature"
+
+  /**
+    * Some initial state that can be used on the very first call to `apply(A, S)`.
+    * @return some state.
+    */
+  @transient override lazy val initialState: Long = {
+
+    val seed = seedCreator()
+
+    // For logging.  Try to get time as close as possible to calling initialSeed.
+    // Note: There's a very good chance this will differ.
+    val time       = System.nanoTime()
+    val ip         = java.net.InetAddress.getLocalHost.getHostAddress
+    val thread     = Thread.currentThread()
+    val threadId   = thread.getId
+    val threadName = thread.getName
+
+    val scrambled  = scramble(seed)
+
+    info(
+      s"${getClass.getSimpleName} seed: $seed, scrambled: $scrambled, " +
+      s"nanotime: $time, ip: $ip, threadId: $threadId, threadName: $threadName"
+    )
+
+    scrambled
+  }
+
+  /**
+    * Given an `a` and some `seed`, produce output, including a new seed.
+    *
+    * When using this function, the user is responsible for keeping track of,
+    * and providing the seeds.
+    *
+    * The implementation of this function should be referentially transparent.
+    *
+    * @param a     input
+    * @param seed the random seed which is updated on each call.
+    * @return a tuple where the first element is a Tuple2 whose first element is
+    *         missing and error information and second element is an optional result.
+    *         The second element of the outer Tuple2 is the new state.
+    */
+  override def apply(a: A, seed: Long): ((MissingAndErroneousFeatureInfo, Option[Array[String]]), Long) = {
+    val (missingAndErrs, features) = featuresFunction(a)
+
+    // Get the lazy val once.
+    val labToInd = labelToInd
+
+    // TODO: This seems like it could be optimized.
```

Comment:
Should this comment be moved to the function called?

Reply:
I don't think so because the computation of `positiveIndices` is "*part of the algorithm*".